File: plugins/data-ai-expert/skills/model-first-reasoning/scripts/validate-model.py
#!/usr/bin/env python3
from __future__ import annotations

"""
Model-First Reasoning Validator

校验 model.json 是否具备实现前必须冻结的结构。
退出码：
- 0：结构有效，可进入实现阶段
- 1：结构无效
- 2：结构有效，但仍存在 unknowns，必须停在 Phase 1
"""

import argparse
import json
from typing import Any

REQUIRED_KEYS = {
    "deliverable",
    "entities",
    "state_variables",
    "actions",
    "constraints",
    "initial_state",
    "goal",
    "assumptions",
    "unknowns",
    "requirement_trace",
    "test_oracles",
}
REQUIRED_ACTION_KEYS = {"name", "preconditions", "effects"}
REQUIRED_CONSTRAINT_KEYS = {"id", "statement"}
REQUIRED_TRACE_KEYS = {"requirement", "represented_as", "ref"}
REQUIRED_ORACLE_KEYS = {"id", "maps_to", "description"}
LIST_FIELDS = [
    "entities",
    "state_variables",
    "actions",
    "constraints",
    "initial_state",
    "goal",
    "assumptions",
    "unknowns",
    "requirement_trace",
    "test_oracles",
]


def validate_list_of_dicts(
    data: dict[str, Any],
    field: str,
    required_keys: set[str],
    label: str,
    issues: list[str],
) -> None:
    """校验列表字段中的每一项是否为对象且包含必需键。"""
    value = data.get(field)
    if not isinstance(value, list):
        return

    for index, item in enumerate(value):
        if not isinstance(item, dict):
            issues.append(f"{label}[{index}] must be an object")
            continue

        missing = required_keys - set(item.keys())
        if missing:
            issues.append(f"{label}[{index}] missing: {', '.join(sorted(missing))}")


def validate_model(data: dict[str, Any]) -> tuple[bool, list[str]]:
    """返回 (is_valid, issues)。"""
    issues: list[str] = []

    missing_keys = REQUIRED_KEYS - set(data.keys())
    if missing_keys:
        issues.append(f"Missing top-level keys: {', '.join(sorted(missing_keys))}")

    deliverable = data.get("deliverable")
    if deliverable is not None:
        if not isinstance(deliverable, dict):
            issues.append("'deliverable' must be an object")
        elif "description" not in deliverable:
            issues.append("'deliverable' missing 'description'")

    for field in LIST_FIELDS:
        value = data.get(field)
        if value is not None and not isinstance(value, list):
            issues.append(f"'{field}' must be a list")

    validate_list_of_dicts(data, "actions", REQUIRED_ACTION_KEYS, "Action", issues)
    validate_list_of_dicts(
        data, "constraints", REQUIRED_CONSTRAINT_KEYS, "Constraint", issues
    )
    validate_list_of_dicts(
        data, "requirement_trace", REQUIRED_TRACE_KEYS, "RequirementTrace", issues
    )
    validate_list_of_dicts(
        data, "test_oracles", REQUIRED_ORACLE_KEYS, "TestOracle", issues
    )

    return len(issues) == 0, issues


def parse_args() -> argparse.Namespace:
    """解析命令行参数。"""
    parser = argparse.ArgumentParser(description="Validate a Model-First Reasoning model file")
    parser.add_argument("model_path", help="Path to model JSON file")
    return parser.parse_args()


def main() -> int:
    """CLI 入口。"""
    args = parse_args()

    try:
        with open(args.model_path, "r", encoding="utf-8") as handle:
            data = json.load(handle)
    except FileNotFoundError:
        print(f"ERROR: File not found: {args.model_path}")
        return 1
    except json.JSONDecodeError as exc:
        print(f"ERROR: Invalid JSON: {exc}")
        return 1

    is_valid, issues = validate_model(data)
    if issues:
        print("VALIDATION FAILED:")
        for issue in issues:
            print(f"  - {issue}")
        print()

    unknowns = data.get("unknowns", [])
    if isinstance(unknowns, list) and unknowns:
        print(f"WARNING: {len(unknowns)} unknowns remain - STOP after Phase 1")
        for unknown in unknowns:
            print(f"  - {unknown}")
        print()
        print("Do NOT proceed to implementation until unknowns are resolved.")
        return 2 if is_valid else 1

    if is_valid:
        print("OK: Model structure is valid")
        print("    Ready for Phase 2: Implementation")
        return 0

    return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: plugins/data-ai-expert/skills/model-first-reasoning/scripts/validate-model.mjs
#!/usr/bin/env node
/**
 * Model-First Reasoning Validator
 *
 * Validates whether model.json contains the structure that must be frozen
 * before implementation.
 *
 * Exit codes:
 * - 0: valid structure, ready for implementation
 * - 1: invalid structure
 * - 2: valid structure, but unknowns remain and Phase 1 must stop
 */

import { existsSync, readFileSync } from "node:fs";

const REQUIRED_KEYS = new Set([
  "deliverable",
  "entities",
  "state_variables",
  "actions",
  "constraints",
  "initial_state",
  "goal",
  "assumptions",
  "unknowns",
  "requirement_trace",
  "test_oracles",
]);
const REQUIRED_ACTION_KEYS = new Set(["name", "preconditions", "effects"]);
const REQUIRED_CONSTRAINT_KEYS = new Set(["id", "statement"]);
const REQUIRED_TRACE_KEYS = new Set(["requirement", "represented_as", "ref"]);
const REQUIRED_ORACLE_KEYS = new Set(["id", "maps_to", "description"]);
const LIST_FIELDS = [
  "entities",
  "state_variables",
  "actions",
  "constraints",
  "initial_state",
  "goal",
  "assumptions",
  "unknowns",
  "requirement_trace",
  "test_oracles",
];

function isObject(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function sortedMissing(requiredKeys, item) {
  return [...requiredKeys].filter((key) => !(key in item)).sort();
}

function validateListOfObjects(data, field, requiredKeys, label, issues) {
  const value = data[field];
  if (!Array.isArray(value)) {
    return;
  }

  value.forEach((item, index) => {
    if (!isObject(item)) {
      issues.push(`${label}[${index}] must be an object`);
      return;
    }

    const missing = sortedMissing(requiredKeys, item);
    if (missing.length > 0) {
      issues.push(`${label}[${index}] missing: ${missing.join(", ")}`);
    }
  });
}

function validateModel(data) {
  const issues = [];

  const missingKeys = [...REQUIRED_KEYS].filter((key) => !(key in data)).sort();
  if (missingKeys.length > 0) {
    issues.push(`Missing top-level keys: ${missingKeys.join(", ")}`);
  }

  const deliverable = data.deliverable;
  if (deliverable !== undefined) {
    if (!isObject(deliverable)) {
      issues.push("'deliverable' must be an object");
    } else if (!("description" in deliverable)) {
      issues.push("'deliverable' missing 'description'");
    }
  }

  for (const field of LIST_FIELDS) {
    const value = data[field];
    if (value !== undefined && !Array.isArray(value)) {
      issues.push(`'${field}' must be a list`);
    }
  }

  validateListOfObjects(data, "actions", REQUIRED_ACTION_KEYS, "Action", issues);
  validateListOfObjects(data, "constraints", REQUIRED_CONSTRAINT_KEYS, "Constraint", issues);
  validateListOfObjects(data, "requirement_trace", REQUIRED_TRACE_KEYS, "RequirementTrace", issues);
  validateListOfObjects(data, "test_oracles", REQUIRED_ORACLE_KEYS, "TestOracle", issues);

  return [issues.length === 0, issues];
}

function usage() {
  return "Usage: node validate-model.mjs <model_path>";
}

function main() {
  const args = process.argv.slice(2);
  if (args.includes("--help") || args.includes("-h")) {
    console.log(usage());
    return 0;
  }

  const modelPath = args[0];
  if (!modelPath) {
    console.error(usage());
    return 1;
  }

  if (!existsSync(modelPath)) {
    console.log(`ERROR: File not found: ${modelPath}`);
    return 1;
  }

  let data;
  try {
    data = JSON.parse(readFileSync(modelPath, "utf8"));
  } catch (error) {
    console.log(`ERROR: Invalid JSON: ${error.message}`);
    return 1;
  }

  const [isValid, issues] = validateModel(data);
  if (issues.length > 0) {
    console.log("VALIDATION FAILED:");
    for (const issue of issues) {
      console.log(`  - ${issue}`);
    }
    console.log("");
  }

  const unknowns = data.unknowns ?? [];
  if (Array.isArray(unknowns) && unknowns.length > 0) {
    console.log(`WARNING: ${unknowns.length} unknowns remain - STOP after Phase 1`);
    for (const unknown of unknowns) {
      console.log(`  - ${unknown}`);
    }
    console.log("");
    console.log("Do NOT proceed to implementation until unknowns are resolved.");
    return isValid ? 2 : 1;
  }

  if (isValid) {
    console.log("OK: Model structure is valid");
    console.log("    Ready for Phase 2: Implementation");
    return 0;
  }

  return 1;
}

process.exitCode = main();
